### PR TITLE
Implement word boundary search and wordwise movement

### PR DESCRIPTION
**Pull request by elkowar**
This PR implements word-boundary search, and uses it to implement word-wise movement.
To make that usable, it also makes the frontend include modifier state when sending key events.

### CODE REVIEW

File: crates/bazed-core/src/input_mapper.rs [hunk 1]
```diff
@@ -2,43 +2,45 @@
...
-    if input.ctrl_held() {
-        match input.key {
-            Key::Char('z') => Some(Operation::Buffer(BufferOp::Undo)),
-            Key::Char('y') => Some(Operation::Buffer(BufferOp::Redo)),
-            Key::Char('s') => Some(Operation::Document(DocumentOp::Save)),
-            _ => None,
-        }
-    } else {
-        use Operation::*;
+    use Operation::*;
```
Comment (buffet) [inline crates/bazed-core/src/input_mapper.rs:11]: asterisk

File: crates/bazed-core/src/word_boundary.rs [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    End,
+    /// Indicates that the position is both a start and an end,
+    /// i.e. in the case of `a__`, at offset 1, as the underscores
+    /// are also grouped into one word.
+    Both,
+}
+
+impl WordBoundaryType {
+    pub(crate) fn between_types(a: CharCategory, b: CharCategory) -> Option<WordBoundaryType> {
+        use CharCategory::*;
```
Comment (buffet) [inline crates/bazed-core/src/word_boundary.rs:50]: asterisk

File: crates/bazed-core/src/word_boundary.rs [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    /// Any character that doesn't fit any other category.
+    /// This also includes format control characters, UTF-16 surrogate code points.
+    Other,
+}
+
+impl CharCategory {
+    fn of_char(c: char) -> Self {
+        if c.is_whitespace() {
+            return Self::Whitespace;
+        }
```
Comment (buffet) [inline crates/bazed-core/src/word_boundary.rs:97]: is this required?

File: crates/bazed-core/src/word_boundary.rs [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+        use WordBoundaryType::*;
+        fn boundaries(s: &str) -> Vec<(usize, WordBoundaryType)> {
+            WordBoundaries::from_iter(false, s.chars()).collect()
+        }
+        let actual = boundaries("foo foo...");
+        assert_eq!(vec![(3, End), (4, Start), (7, Both)], actual);
+        let actual = boundaries(" foo ");
+        assert_eq!(vec![(1, Start), (4, End)], actual);
+
+        // TODO we should have configurable word separators to allow `_` to not break words if the users wants that
```
Comment (buffet) [inline crates/bazed-core/src/word_boundary.rs:223]: Optimally something like `go to end of token` even.

**Review by buffet — CHANGES_REQUESTED** [review]

File: crates/bazed-core/src/word_boundary.rs (hunk 3 above)
Comment (elkowar) [reply crates/bazed-core/src/word_boundary.rs:97]: let's say yes, ish

File: crates/bazed-core/src/word_boundary.rs (hunk 4 above)
Comment (elkowar) [reply crates/bazed-core/src/word_boundary.rs:223]: Ye, that'd be a future wish, I think

File: crates/bazed-core/src/word_boundary.rs (hunk 2 above)
Comment (elkowar) [reply crates/bazed-core/src/word_boundary.rs:50]: closed as wont fix

File: crates/bazed-core/src/input_mapper.rs (hunk 1 above)
Comment (elkowar) [reply crates/bazed-core/src/input_mapper.rs]: fixed

**Review by buffet — APPROVED** [review]

**Pull request merged**